File: boards/grove-vision-ai-v2/test/test_plugin_stack.py
#!/usr/bin/env python3
# SPDX-License-Identifier: MIT
# Copyright (c) 2026 ThreadX Shell Project
"""Host tests for port/npu/nn_plugin_stack.h (issue #119).

WHAT WENT WRONG, AND WHAT THIS HAS TO BE ABLE TO SEE

Until issue #119 this board declared the stack allowance of entry, shapes_ok and
decode against the camera producer's figure -- 4,096 B -- while all three are
also called on a shell thread whose whole stack is 4,096 B.  A plugin declaring
4,096 would have been admitted and overflowed.  Nothing could notice, for two
separate reasons, and this file answers each:

  1. No compile-time check tied an allowance to the threads its slot runs on.
     The header now asserts each slot's allowance below every stack that slot
     runs on, one assert per pair.  An assert nobody has watched fail is not a
     check (issues #42/#66), so every ceiling is driven here at, below and above
     the allowance, WITH THE OTHER CEILINGS OUT OF THE WAY, and the refusal has
     to be EXACTLY the set of asserts this file's own table says run on that
     thread -- one missing is a row deleted from the header, one extra is a
     slot wired to the wrong allowance.  The asserts are the header's: this
     compiles `#include "nn_plugin_stack.h"` and nothing else.  A test that
     restated the comparisons would go on passing with the header's deleted.

  2. Every allowance was 1,024 B, so a slot wired to the wrong one was invisible.
     The runtime half (test_plugin_stack.c) prints the table the firmware builds
     its policy from, compiled with the two allowances set to DIFFERENT sentinel
     values, and it is checked here against this file's own statement of which
     slot runs where.  It also runs the device's validator (svc/plugin_load.c)
     against the REAL allowances: one byte past each limit must be refused.

The real values are read from where the firmware gets them -- board.cmake for
the allowances and the shell stacks, camera.h / cam_lcd_sink.h for the camera
threads -- and each must be found exactly once.  A value this file could not
find is a failure, not a default.

[!] WHAT THIS DOES NOT SEE.  board.cmake states the same slot -> allowance
mapping four more times (the host policy, both plugins' ENTRIES and the compile
definitions); this file checks the firmware's table only.  Nor does it see
nn_svc_grove.c stop using GROVE_PLUGIN_STACK_LIMITS: it tests the header.
"""

import os
import re
import shlex
import subprocess
import sys
import tempfile

HERE = os.path.dirname(os.path.abspath(__file__))
BOARD = os.path.dirname(HERE)
REPO = os.path.dirname(os.path.dirname(BOARD))

INCLUDES = ["-I", os.path.join(BOARD, "port", "npu"),
            "-I", os.path.join(REPO, "shell", "include"),
            "-I", os.path.join(REPO, "svc")]
CFLAGS = shlex.split(os.environ.get("HOST_TEST_CFLAGS", "-std=c11 -Wall -Wextra"))
# -Werror: an "ok" has to mean the header compiled clean, not that a warning
# about one of its comparisons scrolled past.
CFLAGS += ["-Werror"]

# Which threads each slot runs on, and which allowance it is declared against
# -- the table in the header, stated again here ON PURPOSE: this is what the
# header is checked against.
SLOTS = ["entry", "shapes_ok", "decode", "draw", "report", "param_set",
         "param_get"]
THREADS = {"entry": ["console", "bg"], "shapes_ok": ["console", "bg"],
           "decode": ["producer", "console", "bg"], "draw": ["panel"],
           "report": ["console", "bg"], "param_set": ["console", "bg"],
           "param_get": ["console", "bg"]}
EXPECT = {"entry": "SHELL", "shapes_ok": "SHELL", "decode": "SHELL",
          "draw": "PANEL", "report": "SHELL", "param_set": "SHELL",
          "param_get": "SHELL"}

# ceiling -> (the macro that carries it, the allowance its slots take)
CEILINGS = {
    "console":  ("CLI_INSTANCE_STACK_SIZE", "SHELL"),
    "bg":       ("CLI_BG_JOB_STACK_SIZE", "SHELL"),
    "producer": ("CAM_PRODUCER_STACK_BYTES", "SHELL"),
    "panel":    ("CAM_PANEL_STACK_BYTES", "PANEL"),
}


def on(thread):
    """The asserts that must fire when `thread`'s stack is too small."""
    return frozenset("%s < %s" % (s, thread) for s in SLOTS
                     if thread in THREADS[s])


def nonzero(allow):
    """The asserts that must fire when allowance `allow` is 0."""
    return frozenset("%s > 0" % s for s in SLOTS if EXPECT[s] == allow)


def fired(err):
    """Which of the header's asserts refused, by the name each one carries."""
    return frozenset(m.group(1) for ln in err.splitlines()
                     if "static assertion failed" in ln
                     for m in [re.search(r"plugin stack \[([^\]]+)\]", ln)]
                     if m)


def find_one(path, pattern, what):
    """The single value `pattern` captures in `path`, or a failure."""
    with open(path) as fh:
        hits = re.findall(pattern, fh.read(), re.MULTILINE)
    if len(hits) != 1:
        sys.exit("test_plugin_stack: FAILED -- expected exactly one %s in %s, "
                 "found %d (pattern %r)" % (what, os.path.relpath(path, REPO),
                                            len(hits), pattern))
    return int(hits[0])


def real_values():
    cmake = os.path.join(BOARD, "board.cmake")
    cam = os.path.join(BOARD, "port", "camera")
    return {
        "GROVE_PLUGIN_STACK_SHELL": find_one(
            cmake, r"^set\(GROVE_PLUGIN_STACK_SHELL\s+(\d+)\)\s*$",
            "shell allowance"),
        "GROVE_PLUGIN_STACK_PANEL": find_one(
            cmake, r"^set\(GROVE_PLUGIN_STACK_PANEL\s+(\d+)\)\s*$",
            "panel allowance"),
        "CLI_INSTANCE_STACK_SIZE": find_one(
            cmake, r"^\s*CLI_INSTANCE_STACK_SIZE=(\d+)\b", "console stack"),
        "CLI_BG_JOB_STACK_SIZE": find_one(
            cmake, r"^\s*CLI_BG_JOB_STACK_SIZE=(\d+)\b", "job stack"),
        "CAM_PRODUCER_STACK_BYTES": find_one(
            os.path.join(cam, "camera.h"),
            r"^#define\s+CAM_PRODUCER_STACK_BYTES\s+(\d+)u?\s*$",
            "producer stack"),
        "CAM_PANEL_STACK_BYTES": find_one(
            os.path.join(cam, "cam_lcd_sink.h"),
            r"^#define\s+CAM_PANEL_STACK_BYTES\s+(\d+)u?\s*$",
            "panel stack"),
    }


def defines(values):
    return ["-D%s=%su" % (k, v) for k, v in sorted(values.items())
            if v is not None]


def compile_header(work, values, extra=()):
    tu = os.path.join(work, "tu.c")
    with open(tu, "w") as fh:
        fh.write('#include "nn_plugin_stack.h"\n')
    r = subprocess.run(["gcc"] + CFLAGS + ["-fsyntax-only"] + INCLUDES
                       + defines(values) + list(extra) + [tu],
                       capture_output=True, text=True)
    return r.returncode, r.stderr


# --- the compile-time half: each ceiling on its own ------------------------
#
# Two DIFFERENT allowances, so an assert that compared the wrong one is caught
# too: the panel's "equal" case below leaves the shell allowance comfortably
# under it, and the producer's "below" case would trip an assert that compared
# draw's allowance there instead of decode's.
ALLOW = {"SHELL": 1000, "PANEL": 1016}
WIDE = 1 << 16          # a ceiling that is out of the way


def base():
    v = {"GROVE_PLUGIN_STACK_SHELL": ALLOW["SHELL"],
         "GROVE_PLUGIN_STACK_PANEL": ALLOW["PANEL"]}
    for macro, _ in CEILINGS.values():
        v[macro] = WIDE
    return v


def ct_cases(real):
    none = frozenset()
    cases = [("real values", real, [], none, None),
             ("every ceiling wide", base(), [], none, None)]
    for name, (macro, allow) in CEILINGS.items():
        a = ALLOW[allow]
        for label, ceiling, want in (("below", a + 8, none),
                                     ("equal", a, on(name)),
                                     ("above", a - 8, on(name))):
            v = base()
            v[macro] = ceiling
            cases.append(("%s %s (%s %d, allowance %d)"
                          % (name, label, macro, ceiling, a), v, [], want,
                          None))
    for allow in ("SHELL", "PANEL"):
        v = base()
        v["GROVE_PLUGIN_STACK_" + allow] = 0
        cases.append(("%s allowance 0" % allow.lower(), v, [],
                      nonzero(allow), None))
    v = base()
    cases.append(("the retired producer allowance", v,
                  ["-DGROVE_PLUGIN_STACK_PRODUCER=4096u"], None,
                  "was retired by issue #119"))
    v = base()
    v["CAM_PRODUCER_STACK_BYTES"] = None
    cases.append(("a ceiling the includer did not bring", v, [], None,
                  "include camera.h and cam_lcd_sink.h"))
    return cases


def run_ct(real):
    bad = 0
    for what, values, extra, want, must_say in ct_cases(real):
        with tempfile.TemporaryDirectory() as work:
            rc, err = compile_header(work, values, extra)
        names = fired(err)
        # [!] A REFUSAL COUNTS ONLY FOR ITS OWN REASONS.  Any compile error
        # would otherwise "refuse" -- a broken include would pass every
        # negative case while testing nothing.  So a static-assert case must
        # fail with exactly the named asserts and no other error line, and an
        # #error case must say what it was asked to.
        other = [ln for ln in err.splitlines() if "error:" in ln
                 and "static assertion failed" not in ln]
        if must_say is not None:
            ok = rc != 0 and "#error" in err and must_say in err
            said = "refused (#error)" if ok else "not the #error expected"
        elif not want:
            ok = rc == 0
            said = "compiles"
        else:
            ok = rc != 0 and names == want and not other
            said = "refused by exactly %d assert(s)" % len(want)
        if not ok:
            print("  FAIL %-58s" % what)
            if must_say is None:
                print("        wanted %s" % (sorted(want) or "a clean compile"))
                print("        fired  %s" % (sorted(names) or "nothing"))
            for ln in (other or err.strip().splitlines())[:6]:
                print("        " + ln)
            bad += 1
        else:
            print("  ok   %-58s %s" % (what, said))
    return bad


# --- the runtime half: the table, and the validator against it -------------

def run_probe(work, values):
    exe = os.path.join(work, "probe")
    r = subprocess.run(["gcc"] + CFLAGS + INCLUDES + defines(values)
                       + [os.path.join(HERE, "test_plugin_stack.c"),
                          os.path.join(REPO, "svc", "plugin_load.c"),
                          os.path.join(REPO, "svc", "crc32.c"),
                          "-o", exe], capture_output=True, text=True)
    if r.returncode != 0:
        return None, "did not build:\n" + r.stderr
    r = subprocess.run([exe], capture_output=True, text=True)
    table = {}
    for ln in r.stdout.splitlines():
        m = re.match(r"limit (\d+) (\d+)$", ln)
        if m:
            table[SLOTS[int(m.group(1))]] = int(m.group(2))
    return (table if r.returncode == 0 else None), r.stdout


def run_rt(real):
    bad = 0
    # Two sentinel pairs, swapped in order, so a slot that happened to be
    # hard-wired to one of the numbers cannot match both times.
    for label, shell, panel in (("real values", None, None),
                                ("sentinels 1001 / 1013", 1001, 1013),
                                ("sentinels 1019 / 997", 1019, 997)):
        v = dict(real)
        if shell is not None:
            v["GROVE_PLUGIN_STACK_SHELL"] = shell
            v["GROVE_PLUGIN_STACK_PANEL"] = panel
        with tempfile.TemporaryDirectory() as work:
            table, out = run_probe(work, v)
        if table is None:
            print("  FAIL %-58s the validator run failed" % label)
            lines = out.strip().splitlines()
            for ln in [ln for ln in lines if "FAIL" in ln] or lines[:10]:
                print("        " + ln)
            bad += 1
            continue
        refused = sum(1 for ln in out.splitlines() if "PLUGIN_ERR_STACK" in ln
                      or "stack request refused" in ln)
        print("  ok   %-58s one byte past every limit refused (%d refusals)"
              % (label, refused))
        if shell is None:
            continue
        # The mapping, slot by slot, against this file's own table.
        want = {s: v["GROVE_PLUGIN_STACK_" + EXPECT[s]] for s in SLOTS}
        wrong = ["%s=%d (want %s=%d)" % (s, table.get(s, -1), EXPECT[s],
                                         want[s])
                 for s in SLOTS if table.get(s) != want[s]]
        if wrong:
            print("  FAIL %-58s %s" % (label + ": slot -> allowance",
                                       ", ".join(wrong)))
            bad += 1
        else:
            print("  ok   %-58s every slot got the allowance of its threads"
                  % (label + ": slot -> allowance"))
    return bad


def main():
    real = real_values()
    print("test_plugin_stack: allowances shell %d / panel %d; stacks console "
          "%d, job %d, producer %d, panel %d" % (
              real["GROVE_PLUGIN_STACK_SHELL"],
              real["GROVE_PLUGIN_STACK_PANEL"],
              real["CLI_INSTANCE_STACK_SIZE"], real["CLI_BG_JOB_STACK_SIZE"],
              real["CAM_PRODUCER_STACK_BYTES"],
              real["CAM_PANEL_STACK_BYTES"]))
    print("the header's asserts, one ceiling at a time:")
    bad = run_ct(real)
    print("the table the firmware builds its policy from:")
    bad += run_rt(real)
    if bad:
        print("test_plugin_stack: FAILED (%d)" % bad, file=sys.stderr)
        return 1
    print("test_plugin_stack: all cases behave as specified")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: boards/grove-vision-ai-v2/test/test_plugin_stack.c
/*
 * SPDX-License-Identifier: MIT
 * Copyright (c) 2026 ThreadX Shell Project
 *
 * The runtime half of the plugin stack policy test (issue #119).  Built and run
 * by test_plugin_stack.py, several times, each with a different set of -D; do
 * not run it on its own.
 *
 * It does two things with the table port/npu/nn_plugin_stack.h hands the
 * firmware -- the same GROVE_PLUGIN_STACK_LIMITS initialiser nn_svc_grove.c
 * builds nn_plugin_policy from, not a copy of it:
 *
 *   - prints it, one `limit <slot> <bytes>` line per slot, so the driver can
 *     check WHICH allowance each slot got (and `runs <slot> <mask>`, the
 *     threads the header says it runs on).  Compiled with the two allowances set
 *     to different sentinel values, a slot wired to the wrong one shows up as
 *     the wrong number; with the real values (both 1,024) it could not.
 *   - runs svc/plugin_load.c -- the device's validator -- over a container that
 *     declares, slot by slot, exactly the limit (accepted) and one byte more
 *     (PLUGIN_ERR_STACK), and then the whole stack of each thread a slot runs
 *     on: the declaration issue #119 found ACCEPTED for decode.
 *
 * Only the stack table is the firmware's.  The target, prelink base, capacity
 * and veneer cost below are whatever lets the parse reach the stack check; they
 * are not under test here.  Every slot declares its own frames only (no
 * crossing), so what the loader requires of it is exactly what it declares.
 */
#include "nn_plugin_stack.h"
#include "plugin_load.h"
#include "crc32.h"

#include <stdio.h>
#include <string.h>

static const struct plugin_policy pol = {
	.target_id      = 0x9302u,
	.link_addr      = 0x341E0000u,
	.capacity       = 131072u,
	.image_align    = PLUGIN_IMAGE_ALIGN,
	.caps_supported = PLUGIN_CAP_KNOWN_MASK,
	.stack_limit    = GROVE_PLUGIN_STACK_LIMITS,
	.veneer_cost    = 256u,
	.stack_accounting = PLUGIN_STACK_ACCOUNTING,
};

/* ---- a container with all seven slots ------------------------------------ */

#define IMAGE_CODE     256u
#define IMAGE_DATA     64u
#define IMAGE_BSS      128u
#define IMAGE_SCRATCH  1024u
#define IMAGE_FILE     (IMAGE_CODE + IMAGE_DATA)
#define IMAGE_MEM      (IMAGE_FILE + IMAGE_BSS + IMAGE_SCRATCH)   /* 46 x 32 */

#define HDR_SZ         ((uint32_t)sizeof(struct plugin_container_hdr))
#define MAN_SZ         ((uint32_t)sizeof(struct plugin_manifest))
#define SECT_OFF       HDR_SZ
#define SECT_LEN       (MAN_SZ + IMAGE_FILE)
#define ALIGN_UP(v, a) (((v) + (a) - 1u) & ~((a) - 1u))
#define MODEL_OFF      ALIGN_UP(SECT_OFF + SECT_LEN, PLUGIN_MODEL_ALIGN)
#define MODEL_LEN      128u
#define TOTAL          (MODEL_OFF + MODEL_LEN)

static uint8_t buf[TOTAL];

#define HF(f)     ((uint32_t)offsetof(struct plugin_container_hdr, f))
#define SF(i, f)  (HF(sections) + (uint32_t)(i) * \
                   (uint32_t)sizeof(struct plugin_section) + \
                   (uint32_t)offsetof(struct plugin_section, f))
#define MF(f)     (SECT_OFF + (uint32_t)offsetof(struct plugin_manifest, f))

static void wr32(uint32_t at, uint32_t v)
{
	buf[at + 0u] = (uint8_t)(v & 0xFFu);
	buf[at + 1u] = (uint8_t)((v >> 8) & 0xFFu);
	buf[at + 2u] = (uint8_t)((v >> 16) & 0xFFu);
	buf[at + 3u] = (uint8_t)((v >> 24) & 0xFFu);
}

/* Every slot present and declaring 8 B, except the one a case sets. */
static void build(void)
{
	unsigned i;

	memset(buf, 0, sizeof buf);
	buf[0] = PLUGIN_CONTAINER_MAGIC0; buf[1] = PLUGIN_CONTAINER_MAGIC1;
	buf[2] = PLUGIN_CONTAINER_MAGIC2; buf[3] = PLUGIN_CONTAINER_MAGIC3;
	buf[4] = PLUGIN_CONTAINER_FORMAT0; buf[5] = PLUGIN_CONTAINER_FORMAT1;
	buf[6] = PLUGIN_CONTAINER_FORMAT2; buf[7] = PLUGIN_CONTAINER_FORMAT3;
	wr32(HF(hdr_size), HDR_SZ);
	wr32(HF(abi_version), PLUGIN_ABI_VERSION);
	wr32(HF(total_size), TOTAL);
	wr32(HF(section_count), 2u);
	wr32(SF(0, type), PLUGIN_SECTION_PLUGIN);
	wr32(SF(0, offset), SECT_OFF);
	wr32(SF(0, length), SECT_LEN);
	wr32(SF(1, type), PLUGIN_SECTION_MODEL);
	wr32(SF(1, offset), MODEL_OFF);
	wr32(SF(1, length), MODEL_LEN);

	buf[MF(magic) + 0u] = PLUGIN_MANIFEST_MAGIC0;
	buf[MF(magic) + 1u] = PLUGIN_MANIFEST_MAGIC1;
	buf[MF(magic) + 2u] = PLUGIN_MANIFEST_MAGIC2;
	buf[MF(magic) + 3u] = PLUGIN_MANIFEST_MAGIC3;
	wr32(MF(struct_size), MAN_SZ);
	wr32(MF(abi_version), PLUGIN_ABI_VERSION);
	wr32(MF(target_id), pol.target_id);
	wr32(MF(link_addr), pol.link_addr);
	wr32(MF(capability),
	     PLUGIN_CAP_DRAW | PLUGIN_CAP_REPORT | PLUGIN_CAP_PARAMS);
	wr32(MF(image_off), MAN_SZ);
	wr32(MF(file_size), IMAGE_FILE);
	wr32(MF(mem_size), IMAGE_MEM);
	wr32(MF(code_off), 0u);
	wr32(MF(code_len), IMAGE_CODE);
	wr32(MF(data_off), IMAGE_CODE);
	wr32(MF(data_len), IMAGE_DATA);
	wr32(MF(bss_off), IMAGE_FILE);
	wr32(MF(bss_len), IMAGE_BSS);
	wr32(MF(scratch_off), IMAGE_FILE + IMAGE_BSS);
	wr32(MF(scratch_len), IMAGE_SCRATCH);
	for (i = 0u; i < PLUGIN_SLOT_COUNT; i++) {
		wr32(MF(slot) + 4u * i, 0x20u * i + 1u);   /* Thumb bit set */
		wr32(MF(stack_own) + 4u * i, 8u);
	}
	wr32(MF(stack_accounting), PLUGIN_STACK_ACCOUNTING);
}

static enum plugin_result parse_with(unsigned slot, uint32_t declared)
{
	struct plugin_view view;

	build();
	wr32(MF(stack_own) + 4u * slot, declared);
	wr32(HF(plugin_digest), crc32_update(0u, buf + SECT_OFF, SECT_LEN));
	return plugin_parse(buf, sizeof buf, &pol, &view);
}

static int failures;

static void expect(unsigned slot, const char *what, uint32_t declared,
                   enum plugin_result want)
{
	enum plugin_result got = parse_with(slot, declared);

	if (got != want) {
		printf("  FAIL slot %u %-26s %5lu B -> %s, want %s\n", slot, what,
		       (unsigned long)declared, plugin_result_name(got),
		       plugin_result_name(want));
		failures++;
	} else {
		printf("  ok   slot %u %-26s %5lu B -> %s\n", slot, what,
		       (unsigned long)declared, plugin_result_name(got));
	}
}

int main(void)
{
	/* The stack of every thread each slot can run on (nn_plugin_stack.h). */
	static const uint32_t shell[] = { CLI_INSTANCE_STACK_SIZE,
	                                  CLI_BG_JOB_STACK_SIZE };
	unsigned i, k;

	for (i = 0u; i < PLUGIN_SLOT_COUNT; i++)
		printf("limit %u %lu\n", i, (unsigned long)pol.stack_limit[i]);
	/* And which threads the header says each slot runs on -- the masks the
	 * asserts are gated by and the stack report marks coverage with. */
	{
		static const unsigned runs[PLUGIN_SLOT_COUNT] = GROVE_PLUGIN_STACK_RUNS;

		for (i = 0u; i < PLUGIN_SLOT_COUNT; i++)
			printf("runs %u %u\n", i, runs[i]);
	}

	for (i = 0u; i < PLUGIN_SLOT_COUNT; i++) {
		uint32_t lim = pol.stack_limit[i];

		expect(i, "declares the limit", lim, PLUGIN_OK);
		expect(i, "declares one byte more", lim + 1u, PLUGIN_ERR_STACK);
		if (i == PLUGIN_SLOT_DRAW) {
			expect(i, "declares the panel stack",
			       CAM_PANEL_STACK_BYTES, PLUGIN_ERR_STACK);
			continue;
		}
		for (k = 0u; k < 2u; k++)
			expect(i, k == 0u ? "declares a console stack"
			                  : "declares a job stack",
			       shell[k], PLUGIN_ERR_STACK);
		if (i == PLUGIN_SLOT_DECODE)
			expect(i, "declares the producer stack",
			       CAM_PRODUCER_STACK_BYTES, PLUGIN_ERR_STACK);
	}

	if (failures) {
		printf("test_plugin_stack: %d FAILED\n", failures);
		return 1;
	}
	printf("test_plugin_stack: the policy refuses one byte past every limit\n");
	return 0;
}
